File: src/org/webfabric/web/sitemesh/DivCapturingRule.java
package org.webfabric.web.sitemesh;

import com.opensymphony.module.sitemesh.html.BasicRule;
import com.opensymphony.module.sitemesh.html.Tag;
import com.opensymphony.module.sitemesh.html.rules.PageBuilder;
import com.opensymphony.module.sitemesh.html.util.CharArray;

import java.util.Stack;


public class DivCapturingRule extends BasicRule {
    private final PageBuilder pageBuilder;
    private Stack<String> ids = new Stack<String>();

    public DivCapturingRule(PageBuilder pageBuilder) {
        super("div");
        this.pageBuilder = pageBuilder;
    }

    public void process(Tag tag) {
        if (tag.getType() == Tag.OPEN) {
            ensureTagIsNotConsumed(tag);
            if(shouldCapture(tag)){
               captureContent();
            }
            pushId(tag);
        } else if (tag.getType() == Tag.CLOSE) {
            String id = popId();
            if(capturing(id)){
                String content = popContent();
                pageBuilder.addProperty("div." + id, content);
                ensureContentIsNotConsumed(content);
            }
            ensureTagIsNotConsumed(tag);
        }
    }

    private void ensureContentIsNotConsumed(String content) {
        currentBuffer().append(content);
    }

    private String popContent() {
        return context.popBuffer().toString();
    }

    private boolean capturing(String id) {
        return id != null;
    }

    private String popId() {
        return ids.pop();
    }

    private void pushId(Tag tag) {
        ids.push(tag.getAttributeValue("id", false));
    }

    private void captureContent() {
        context.pushBuffer(new CharArray(256));
    }

    private boolean shouldCapture(Tag tag) {
        return tag.hasAttribute("id", false);
    }

    private void ensureTagIsNotConsumed(Tag tag) {
        tag.writeTo(currentBuffer());
    }


}

File: src/org/webfabric/web/sitemesh/DivCapturingRule.scala
package org.webfabric.web.sitemesh

import com.opensymphony.module.sitemesh.html.rules.PageBuilder
import com.opensymphony.module.sitemesh.html.util.CharArray
import com.opensymphony.module.sitemesh.html.{BasicRule, Tag}
import java.util.Stack

class DivCapturingRule(pageBuilder: PageBuilder) extends BasicRule("div") {
  val ids: Stack[String] = new Stack[String]

  def process(tag: Tag):Unit = tag.getType match {
    case Tag.OPEN => {
      ensureTagIsNotConsumed(tag)
      if (shouldCapture(tag)) {
        pushContent
      }
      pushId(tag)
    }
    case Tag.CLOSE => {
      val id: String = popId
      if (capturing(id)) {
        val content = popContent
        pageBuilder.addProperty("div." + id, content)
        ensureContentIsNotConsumed(content)
      }
      ensureTagIsNotConsumed(tag)
    }
  }

  def ensureContentIsNotConsumed(content: String) {
    currentBuffer().append(content)
  }

  def popContent: String = {
    context.popBuffer().toString();
  }

  def capturing(id: String): Boolean = {
    id != null
  }

  def popId: String = {
    ids.pop()
  }

  def pushId(tag: Tag) {
    ids.push(tag.getAttributeValue("id", false))
  }

  def pushContent {
    context.pushBuffer(new CharArray(256))
  }

  def shouldCapture(tag: Tag): Boolean = {
    return tag.hasAttribute("id", false)
  }

  def ensureTagIsNotConsumed(tag: Tag) {
    tag.writeTo(currentBuffer())
  }
}
